packages/ragbits-core/src/ragbits/core/utils/dict_transformations.py

```diff
@@ -0,0 +1,145 @@
+    """
+    items: dict[str, str | int | float | bool] = {}
+    for k, v in input_dict.items():
+        if sep in k:
+            raise ValueError(f"Separator '{sep}' found in key '{parent_key}' Cannot flatten dictionary safely.")
+
+        if k.endswith("]"):
+            raise ValueError(f"Key '{k}' cannot end with ']' Cannot flatten dictionary safely.")
+
+        new_key = f"{parent_key}{sep}{k}" if parent_key else k
+
+        if isinstance(v, dict):
+            items = {**items, **flatten_dict(v, new_key, sep=sep)}
+        elif isinstance(v, list):
+            for i, item in enumerate(v):
+                list_key = f"{new_key}[{i}]"
+                if isinstance(item, dict):
+                    items = {**items, **flatten_dict(item, list_key, sep=sep)}
+                else:
+                    items[list_key] = item if isinstance(item, str | int | float | bool) else str(item)
+        else:
+            items[new_key] = v if isinstance(v, str | int | float | bool) else str(v)
+
+    return items
+
+
+def unflatten_dict(input_dict: dict[Any, Any]) -> dict[Any, Any] | list:
+    """
+    Converts a flattened dictionary with dot notation and array notation into a nested structure.
+
+    This function transforms a dictionary with flattened keys (using dot notation for nested objects
+    and bracket notation for arrays) into a nested dictionary or list structure. It handles both
+    object-like nesting (using dots) and array-like nesting (using brackets).
+
+    Args:
+        input_dict (dict[Any, Any]): A dictionary with flattened keys. Keys can use dot notation
+            (e.g., "person.name") or array notation (e.g., "addresses[0].street").
+
+    Returns:
+        Union[dict[Any, Any], list]: A nested dictionary or list structure. Returns a list if all
+        top-level keys are consecutive integer strings starting from 0.
+
+    Examples:
+        >>> unflatten_dict({"person.name": "John", "person.age": 30})
+        {'person': {'name': 'John', 'age': 30}}
+
+        >>> unflatten_dict({"addresses[0].street": "Main St", "addresses[1].street": "Broadway"})
+        {'addresses': [{'street': 'Main St'}, {'street': 'Broadway'}]}
+
+        >>> unflatten_dict({"0": "first", "1": "second"})
+        ['first', 'second']
+
+    Notes:
+        - The function recursively processes nested structures
+        - If all keys at any level are consecutive integers starting from 0, that level will be
+          converted to a list
+        - The function preserves the original values for non-nested keys
+        - Keys are sorted before processing to ensure consistent results
+
+    Attribution:
+        - This function is based on the answer by user "djtubig-malicex" on Stack Overflow: https://stackoverflow.com/a/67905359/27947364
+    """
+    new_dict: dict[Any, Any] = {}
+    field_keys = sorted(input_dict.keys())
+
+    def _parse_key(key: str) -> tuple[str | int | None, str | int | None]:
```

Comment:
Returning `None, None` for a simple, flat key (e.g. `"key1"`) is an unusual design choice for a function called `_parse_key()`

Reply:
yeah, not a biggest fan of this, but it covers all the corner-cases and hopefully won't be touched ever again :sweat_smile: 

Renamed to _decompose_key which imo makes more sense in that scenario